web/src/stream_topic_history.ts

```diff
@@ -279,6 +348,120 @@ export function add_message(opts: {
     history.add_or_update(topic_name, message_id, is_data_reliable);
 }
 
+export function add_moved_message(opts: {
+    stream_id: number;
+    max_moved_message_id: number;
+    topic_name: string;
+    can_latest_message_be_inaccessible: boolean;
+}): void {
+    const stream_id = opts.stream_id;
+    const history = find_or_create(stream_id);
+
+    const topic_name = opts.topic_name;
+    const existing_topic = history.topics.get(topic_name);
+    const max_moved_message_id = opts.max_moved_message_id;
+
+    if (existing_topic && existing_topic.message_id > max_moved_message_id) {
+        // In this case, the topic order will not change irrespective
+        // of whether there can be a case where message_id is inaccessible
+        // or the data we already have is reliable or not. So we just
+        // skip.
+        return;
+    }
+
+    if (!opts.can_latest_message_be_inaccessible) {
+        // We are certain that the value of message_id field
+        // is reliable, i.e. we can access the message.
+        // We would update the data in two cases -
+        // 1. Topic with same name exists already.
+        // 2. We are sure that there is no existing topic with
+        // same name in this stream.
+        if (existing_topic !== undefined || is_complete_for_stream_id(stream_id)) {
+            const is_data_reliable = existing_topic ? existing_topic.is_data_reliable : true;
```

Comment:
`existing_topic !== undefined` already, so this can be simplified.

Reply:
I do not understand this clearly. `existing_topic` can be `undefined` as well as we use `||` condition.